src/utils/logger.rs

```diff
@@ -0,0 +1,197 @@
+                } else {
+                    Some(format!("{}[WARNING]{} {}", *ORANGE, *RESET, self.contents))
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum MessageLevel {
+    Info,
+    Warning,
+}
+
+mod logger_thread {
+    use std::{
+        sync::{mpsc::RecvTimeoutError, Arc, Barrier},
+        time::Duration,
+    };
+
+    use super::*;
+
+    type LogReceiver = mpsc::Receiver<Option<PrintMessage>>;
+    type LogSender = mpsc::Sender<Option<PrintMessage>>;
+
+    static SENDER: OnceLock<LogSender> = OnceLock::new();
+
+    #[track_caller]
+    fn setup_channel() -> LogReceiver {
+        let (tx, rx) = mpsc::channel();
+        SENDER.set(tx).expect("`setup_channel` should only be called once");
+        rx
+    }
+
+    #[track_caller]
+    fn get_sender() -> &'static LogSender {
+        SENDER.get().expect("No sender, you need to call `setup_channel` first")
+    }
+
+    #[track_caller]
+    pub(super) fn send_log_message(msg: PrintMessage) {
+        send_message(Some(msg));
+    }
+
+    #[track_caller]
+    fn send_message(msg: Option<PrintMessage>) {
+        get_sender().send(msg).expect("Failed to send internal message");
+    }
+
+    pub struct LoggerThreadHandle {
+        shutdown_barrier: Arc<Barrier>,
+    }
+
+    impl LoggerThreadHandle {
+        /// Tell logger to shutdown and waits till it does.
+        pub fn shutdown_and_wait(self) {
+            // Signal the shutdown
+            send_message(None);
+            // Wait for confirmation
+            self.shutdown_barrier.wait();
+        }
+    }
+
+    pub fn spawn_logger_thread() -> LoggerThreadHandle {
+        let log_receiver = setup_channel();
+
+        let shutdown_barrier = Arc::new(Barrier::new(2));
+
+        let handle = LoggerThreadHandle {
+            shutdown_barrier: shutdown_barrier.clone(),
+        };
+
+        rayon::spawn(move || run_logger(log_receiver, shutdown_barrier));
+
+        handle
+    }
+
+    fn run_logger(log_receiver: LogReceiver, shutdown_barrier: Arc<Barrier>) {
+        const FLUSH_TIMEOUT: Duration = Duration::from_millis(250);
+
+        let mut buffer = Vec::<String>::with_capacity(16);
+
+        loop {
+            let msg = match log_receiver.recv_timeout(FLUSH_TIMEOUT) {
+                Ok(msg) => msg,
+                Err(RecvTimeoutError::Timeout) => {
+                    flush_logs_to_stderr(&mut buffer);
+                    continue;
+                }
+                Err(RecvTimeoutError::Disconnected) => unreachable!("sender is static"),
+            };
+
+            let is_shutdown_message = msg.is_none();
+
+            // Append message to buffer
+            if let Some(msg) = msg.as_ref().and_then(PrintMessage::to_processed_message) {
+                buffer.push(msg);
+            }
+
+            let should_flush = buffer.len() == buffer.capacity() || is_shutdown_message;
```

Comment:
This should probably be

```rs
let should_flush = buffer.len() >= buffer.capacity() || is_shutdown_message;
```

to account for cases where more messages are sent while the lock is held.

Reply:
Actually on second thoughts not sure this is needed since no new messages will be pushed since the loop won't continue until it's flushed right? leaving this here just to be sure im not wrong, if it's fine as is I think everything is good

Reply:
Yeah exactly, but we can change to `>=` too cause why not :) .